File: scripts/hygiene/derivation-sync.py
#!/usr/bin/env python3
# make derivation-sync
# #37's own sync test (DROPPED/UNDECLARED/UNRECORDED/VERSION-MISMATCH
# verdicts).
import json
import os
import sys

import yaml

args = sys.argv[1:]
root = args[args.index("--root") + 1] if "--root" in args else os.getcwd()

DERIVED_DIR = os.path.join(root, "test", "fixtures", "derived")
MANIFEST_PATH = os.path.join(DERIVED_DIR, "manifest.json")
ROSTER_PATH = os.path.join(root, "test", "fixtures", "roster.json")


def list_yaml_files(dir_):
    if not os.path.isdir(dir_):
        return []
    out = []

    def walk(d):
        for entry in sorted(os.listdir(d)):
            full = os.path.join(d, entry)
            if os.path.isdir(full):
                walk(full)
            elif entry.endswith(".yaml") or entry.endswith(".yml"):
                out.append(full)

    walk(dir_)
    return out


def fail(violations):
    print("derivation-sync: FAIL", file=sys.stderr)
    for v in violations:
        print(f"  {v}", file=sys.stderr)
    sys.exit(1)


if not os.path.exists(MANIFEST_PATH):
    fail([f"{os.path.relpath(MANIFEST_PATH, root)} does not exist — nothing records what the derivation repository dropped here"])

try:
    with open(MANIFEST_PATH, "r", encoding="utf-8") as f:
        manifest = json.load(f)
except Exception as e:
    fail([f"{os.path.relpath(MANIFEST_PATH, root)}: could not parse JSON ({e})"])

structural = []
if not isinstance(manifest.get("tool"), str) or len(manifest.get("tool", "")) == 0:
    structural.append("manifest.tool must be a non-empty string")
if not isinstance(manifest.get("tool_version"), str) or len(manifest.get("tool_version", "")) == 0:
    structural.append("manifest.tool_version must be a non-empty string")
if not isinstance(manifest.get("fixtures"), list):
    structural.append("manifest.fixtures must be an array of fixture identities")
if structural:
    fail(structural)

expected_derived_by = f"{manifest['tool']}@{manifest['tool_version']}"
manifest_set = set(manifest["fixtures"])

violations = []
actual_ids = set()

for file in list_yaml_files(DERIVED_DIR):
    rel = os.path.relpath(file, root)
    try:
        with open(file, "r", encoding="utf-8") as f:
            doc = yaml.safe_load(f) or {}
    except Exception as e:
        violations.append(f"{rel}: could not parse YAML ({e})")
        continue
    if not doc.get("fixture"):
        violations.append(f'{rel}: missing required "fixture" identity field')
        continue
    actual_ids.add(doc["fixture"])

    provenance = doc.get("provenance") or {}
    derived_by = provenance.get("derived_by") if isinstance(provenance, dict) else None
    if derived_by != expected_derived_by:
        violations.append(
            f'VERSION-MISMATCH: {rel} ("{doc["fixture"]}") has provenance.derived_by {json.dumps(derived_by)}, expected "{expected_derived_by}" (manifest.tool@manifest.tool_version)'
        )

for id_ in manifest_set:
    if id_ not in actual_ids:
        violations.append(f'DROPPED: manifest declares "{id_}" but no test/fixtures/derived/**/*.yaml fixture carries that identity')
for id_ in actual_ids:
    if id_ not in manifest_set:
        violations.append(f'UNDECLARED: "{id_}" exists under test/fixtures/derived but the manifest never declared it')

if os.path.exists(ROSTER_PATH):
    with open(ROSTER_PATH, "r", encoding="utf-8") as f:
        roster = json.load(f)
else:
    roster = []
roster_set = set(roster)
for id_ in manifest_set:
    if id_ not in roster_set:
        violations.append(f'UNRECORDED: "{id_}" is in the derivation manifest but not yet in {os.path.relpath(ROSTER_PATH, root)}')

if violations:
    fail(violations)
print(f"derivation-sync: PASS ({len(manifest_set)} manifest fixture(s), tool {expected_derived_by})")

File: scripts/hygiene/derivation-sync.mjs
#!/usr/bin/env bun
// make derivation-sync
// #37's own sync test: "test/fixtures/derived/ receives files from the
// derivation repository; a sync test compares the derivation manifest
// against the roster and flags UNRECORDED; the tool's version is in every
// provenance" (issue #37's Goal) — the same UNRECORDED vocabulary
// docs/testing/corpora.md's own promqltest row already names for this exact
// drop-in pipeline: "A derivation tool parses the scripts, isolates
// single-function range evals, and emits neutral fixtures with provenance
// prometheus-<version>; a sync test on the kernel side treats new fixtures
// as UNRECORDED."
//
// The comparison is against `test/fixtures/derived/manifest.json` — the
// derivation tool's own declared inventory of what it emitted this batch —
// never against the tree's actual file list alone, because a plain file
// count can't catch the failure mode corpora.md's own closing line names:
// "A derivation tool that silently drops evals it cannot parse would shrink
// the kernel's corpus without anything going red." A manifest that always
// agreed with whatever files happen to be present could never catch that;
// this script treats the manifest as the source of truth to check the
// files against, not the other way around.
//
// Four fatal verdicts:
//   - DROPPED          the manifest names a fixture identity no file under
//                       test/fixtures/derived/**/*.yaml actually declares
//                       (the corpus shrank silently, corpora.md's own
//                       named risk).
//   - UNDECLARED        a file exists under test/fixtures/derived/**/*.yaml
//                       whose `fixture:` identity the manifest never named
//                       (an ingest that bypassed the tool's own inventory).
//   - UNRECORDED        the manifest names a fixture identity absent from
//                       `test/fixtures/roster.json` — the same verdict name
//                       `kernel-fixture-loader.mjs` already uses for "passes
//                       but is unreviewed onto the roster," checked here
//                       against the tool's declared set directly, so it
//                       fires even before the loader ever compiles
//                       anything.
//   - VERSION-MISMATCH  a fixture's `provenance.derived_by` is not exactly
//                       `<manifest.tool>@<manifest.tool_version>` — the
//                       Goal's flat "the tool's version is in every
//                       provenance" rule, made mechanically checkable
//                       rather than left as an unenforced convention.
import { readFileSync, readdirSync, statSync, existsSync } from "node:fs";
import { join, relative } from "node:path";
import { parse } from "yaml";

const args = process.argv.slice(2);
const rootIdx = args.indexOf("--root");
const root = rootIdx === -1 ? process.cwd() : args[rootIdx + 1];

const DERIVED_DIR = join(root, "test", "fixtures", "derived");
const MANIFEST_PATH = join(DERIVED_DIR, "manifest.json");
const ROSTER_PATH = join(root, "test", "fixtures", "roster.json");

function listYamlFiles(dir) {
  if (!existsSync(dir)) return [];
  const out = [];
  const walk = (d) => {
    for (const entry of readdirSync(d).sort()) {
      const full = join(d, entry);
      const st = statSync(full);
      if (st.isDirectory()) walk(full);
      else if (entry.endsWith(".yaml") || entry.endsWith(".yml")) out.push(full);
    }
  };
  walk(dir);
  return out;
}

function fail(violations) {
  console.error("derivation-sync: FAIL");
  for (const v of violations) console.error(`  ${v}`);
  process.exit(1);
}

if (!existsSync(MANIFEST_PATH)) {
  fail([`${relative(root, MANIFEST_PATH)} does not exist — nothing records what the derivation repository dropped here`]);
}

let manifest;
try {
  manifest = JSON.parse(readFileSync(MANIFEST_PATH, "utf8"));
} catch (e) {
  fail([`${relative(root, MANIFEST_PATH)}: could not parse JSON (${e.message})`]);
}

const structural = [];
if (typeof manifest.tool !== "string" || manifest.tool.length === 0) {
  structural.push(`manifest.tool must be a non-empty string`);
}
if (typeof manifest.tool_version !== "string" || manifest.tool_version.length === 0) {
  structural.push(`manifest.tool_version must be a non-empty string`);
}
if (!Array.isArray(manifest.fixtures)) {
  structural.push(`manifest.fixtures must be an array of fixture identities`);
}
if (structural.length > 0) fail(structural);

// The Goal's own rule, made a single string every fixture's own
// `provenance.derived_by` must match exactly — see docs/testing/registry-and-fixtures.md's
// fixture format, whose own example spells this same shape `derived_by:
// upstream-script-derive@1.4`.
const expectedDerivedBy = `${manifest.tool}@${manifest.tool_version}`;
const manifestSet = new Set(manifest.fixtures);

const violations = [];
const actualIds = new Set();

// Same "read the fixture's own `fixture:` field, never its file name"
// posture `kernel-fixture-loader.mjs` already established for identity.
for (const file of listYamlFiles(DERIVED_DIR)) {
  const rel = relative(root, file);
  let doc;
  try {
    doc = parse(readFileSync(file, "utf8")) ?? {};
  } catch (e) {
    violations.push(`${rel}: could not parse YAML (${e.message})`);
    continue;
  }
  if (!doc.fixture) {
    violations.push(`${rel}: missing required "fixture" identity field`);
    continue;
  }
  actualIds.add(doc.fixture);

  const derivedBy = doc.provenance && doc.provenance.derived_by;
  if (derivedBy !== expectedDerivedBy) {
    violations.push(
      `VERSION-MISMATCH: ${rel} ("${doc.fixture}") has provenance.derived_by ${JSON.stringify(derivedBy ?? null)}, expected "${expectedDerivedBy}" (manifest.tool@manifest.tool_version)`,
    );
  }
}

for (const id of manifestSet) {
  if (!actualIds.has(id)) {
    violations.push(`DROPPED: manifest declares "${id}" but no test/fixtures/derived/**/*.yaml fixture carries that identity`);
  }
}
for (const id of actualIds) {
  if (!manifestSet.has(id)) {
    violations.push(`UNDECLARED: "${id}" exists under test/fixtures/derived but the manifest never declared it`);
  }
}

const roster = existsSync(ROSTER_PATH) ? JSON.parse(readFileSync(ROSTER_PATH, "utf8")) : [];
const rosterSet = new Set(roster);
for (const id of manifestSet) {
  if (!rosterSet.has(id)) {
    violations.push(`UNRECORDED: "${id}" is in the derivation manifest but not yet in ${relative(root, ROSTER_PATH)}`);
  }
}

if (violations.length > 0) fail(violations);
console.log(`derivation-sync: PASS (${manifestSet.size} manifest fixture(s), tool ${expectedDerivedBy})`);
